File: mcp-common/src/main/java/com/mcbbs/mcp/common/McpWebSocketClient.java
package com.mcbbs.mcp.common;

import org.java_websocket.client.WebSocketClient;
import org.java_websocket.handshake.ServerHandshake;
import java.net.URI;
import java.util.concurrent.ConcurrentLinkedQueue;

public class McpWebSocketClient extends WebSocketClient {

    private final ConcurrentLinkedQueue<String> messageQueue = new ConcurrentLinkedQueue<>();
    protected final McpMessageHandler handler;

    public McpWebSocketClient(String serverUrl, McpMessageHandler handler) {
        super(URI.create(serverUrl));
        this.handler = handler;
        this.setConnectionLostTimeout(0);
    }

    public void connectAsync() {
        new Thread(() -> {
            try {
                connect();
            } catch (Exception e) {
                System.err.println("[MCP-WS] Connect failed: " + e.getMessage());
            }
        }, "MCP-WS-Connect").start();
    }

    @Override
    public void onOpen(ServerHandshake handshake) {
        System.out.println("[MCP-WS] Connected to MCP server");
        handler.sendInit(this);
    }

    @Override
    public void onMessage(String message) {
        if (message != null) {
            messageQueue.add(message);
        }
    }

    @Override
    public void onClose(int code, String reason, boolean remote) {
        System.out.println("[MCP-WS] Disconnected: " + reason + " (code=" + code + ")");
    }

    @Override
    public void onError(Exception ex) {
        System.err.println("[MCP-WS] Error: " + (ex != null ? ex.getMessage() : "null"));
    }

    @Override
    public void send(String msg) {
        if (isOpen()) {
            super.send(msg);
        }
    }

    public void handleMessages() {
        String msg;
        while ((msg = messageQueue.poll()) != null) {
            try {
                handler.handleMessage(msg, this);
            } catch (Exception e) {
                System.err.println("[MCP-WS] Handle error: " + e.getMessage());
            }
        }
    }
}

File: mcp-common/src/main/kotlin/com/mcbbs/mcp/common/McpWebSocketClient.kt
package com.mcbbs.mcp.common

import org.java_websocket.client.WebSocketClient
import org.java_websocket.handshake.ServerHandshake
import java.net.URI

open class McpWebSocketClient(
    serverUrl: String,
    protected val handler: McpMessageHandler
) : WebSocketClient(URI.create(serverUrl)) {
    protected var connected = false
        private set

    init { connectionLostTimeout = 0 }

    fun connectAsync() {
        Thread({
            try { connect() } catch (e: Exception) {
                System.err.println("[MCP-WS] Connect failed: ${e.message}")
            }
        }, "MCP-WS-Connect").start()
    }

    override fun onOpen(handshake: ServerHandshake?) {
        connected = true
        System.out.println("[MCP-WS] Connected to MCP server")
        handler.sendInit(this)
    }

    override fun onMessage(message: String?) {
        if (message != null) handler.handleMessage(message, this)
    }

    override fun onClose(code: Int, reason: String?, remote: Boolean) {
        connected = false
        System.out.println("[MCP-WS] Disconnected: $reason (code=$code)")
    }

    override fun onError(ex: Exception?) {
        System.err.println("[MCP-WS] Error: ${ex?.message}")
    }

    override fun send(msg: String) {
        if (connected && isOpen) super.send(msg)
    }

    open fun handleMessages() {}
}
